### PR TITLE
Documentation: overview of java concurrency instrumentation

### PR DESCRIPTION
**Pull request by Degiorgio**

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: f0424d958e6b4659d7ff3e81a58af4cff227d7ae).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91267847

File: jbmc/src/java_bytecode/README.md [hunk 1]
```diff
@@ -140,7 +140,186 @@ To be documented.
 
 \section java-bytecode-concurrency-instrumentation Concurrency instrumentation
 
-To be documented.
+Relevant code:
+
+  - java_bytecode_concurrency_instrumentation.h
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/README.md]: Make these links, please

File: jbmc/src/java_bytecode/README.md [hunk 2]
```diff
@@ -140,7 +140,186 @@ To be documented.
...
+  - java_bytecode_concurrency_instrumentation.cpp
+  - java_static_initializers.h
+  - java_static_initializers.cpp
+  - java_bytecode_convert_method.cpp
+
+Concurrency instrumentation for java programs is disabled by default. It can
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/README.md]: Java (several occurrences)

File: jbmc/src/java_bytecode/README.md [hunk 3]
```diff
@@ -140,7 +140,186 @@ To be documented.
...
+be enabled by specifying the '--java-threading' command line flag.
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/README.md]: Use backticks (several occurrences)

File: jbmc/src/java_bytecode/README.md [hunk 4]
```diff
@@ -140,7 +140,186 @@ To be documented.
...
+
+The instrumentation process (described here) will transform the aforementioned
+function calls, synchronized blocks/methods and calls to
+'java.lang.Thread.getCurrentThreadID:()I' into appropriate codet. As part of
+this process a new variable per thread  is created  along with a single global
+variable that is used keep track of thread identifiers. These
+variables are are needed to handle calls to
+'java.lang.Thread.getCurrentThreadID:()I'.
+
+Hold on, how do we go from java threads to CProver.startThread(I)V and
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/README.md]: \`\`

File: jbmc/src/java_bytecode/README.md [hunk 5]
```diff
@@ -140,7 +140,186 @@ To be documented.
...
+CProver.endThread:(I)V? Well, two ways
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/README.md]: \`\`

File: jbmc/src/java_bytecode/README.md [hunk 6]
```diff
@@ -140,7 +140,186 @@ To be documented.
...
+
+  - You are free to use the CProver API and manually and insert calls to the
+    CProver.startThread() and CProver.EndThread().
+  - Models! Which provide a mock implementation of the 'java.lang.Thread'
+    class.
+
+Note: instrumentation of java static initializers changes when the
+'--java-threading' flag is specified, this is because the static initialiser of
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/README.md]: initializer (to be consistent)

File: jbmc/src/java_bytecode/README.md [hunk 7]
```diff
@@ -140,7 +140,186 @@ To be documented.
...
+every class needs to be carefully synchronized to prevent superfluous
+interleavings.
+
+Note': conversation of synchronised methods (void synchronized test(){..}) and
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/README.md]: synchronized (to be consistent)

File: jbmc/src/java_bytecode/README.md [hunk 8]
```diff
@@ -140,7 +140,186 @@ To be documented.
...
+thread-blocks.  Specifically, function calls to 'CProver.startThread:(I)V' are
+transformed into a codet(id=ID_start_thread, destination=I), which carries a target label in
+the field 'destination'. Similarly 'CProver.endThread(I)V' is transformed into
+a codet(id=ID_end_thread).
+
+For each new thread a thread local variable '__CPROVER__thread_id' is created.
+The new id is obtained by incrementing a global variable
+'__CPROVER__next_thread_id'.
+
+The semantics of codet(id=ID_start_thread, destination=I) roughly corresponds to: fork the
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/README.md]: fork -> spawn

**Comment by Degiorgio** [comment]
@peterschrammel  done

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 855b20d244d1c0adb74e27b8d947a45f37ba2ea9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91358309

**Review by peterschrammel — APPROVED** [review]
Is a good start.

**Pull request merged**